lsp-positions/src/lib.rs

```diff
@@ -0,0 +1,351 @@
+// -*- coding: utf-8 -*-
+// ------------------------------------------------------------------------------------------------
+// Copyright © 2021, stack-graphs authors.
+// Licensed under either of Apache License, Version 2.0, or MIT license, at your option.
+// Please see the LICENSE-APACHE or LICENSE-MIT files in this distribution for license details.
+// ------------------------------------------------------------------------------------------------
+
+//! Defines LSP-compatible positioning information for source code.
+//!
+//! When writing a tool that analyzes or operates on source code, there's a good chance you need to
+//! interoperate with the [Language Server Protocol][lsp].  This seemingly simple requirement makes
+//! it surprisingly difficult to deal with _character locations_.  This is because Rust wants to
+//! store Unicode string content (i.e., the source code you're analyzing) in UTF-8, while
+//! LSP wants to specify character locations using [_UTF-16 code points_][lsp-utf16].
+//!
+//! That means that we ideally need to keep track of each source code position using at least two
+//! units: the UTF-8 offset within the file or containing line (to make it easy to index into UTF-8
+//! encoded strings), as well as the UTF-16 code point offset within the line (to make it possible
+//! to generate `Position` values for LSP).
+//!
+//! [lsp]: https://microsoft.github.io/language-server-protocol/
+//! [lsp-utf16]: https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocuments
+
+use std::ops::Range;
+
+use memchr::memchr;
+
+fn utf16_len(string: &str) -> usize {
+    string.chars().map(char::len_utf16).sum()
+}
+
+/// All of the position information that we have about a character in a source file
+#[repr(C)]
+#[derive(Clone, Debug, Default, Eq, Hash, PartialEq)]
+pub struct Position {
+    /// The 0-indexed line number containing the character
+    pub line: usize,
+    /// The offset of the character within its containing line
+    pub column: Offset,
+    /// The byte offset (within the file) of the line containing the character
+    pub containing_line: Range<usize>,
+    /// The byte offset (within the file) of the line containing the character, with any leading
+    /// and trailing whitespace removed
+    pub trimmed_line: Range<usize>,
```

Comment:
Comments talk about offsets, while types are ranges.

Reply:
Updated the comments here, let me know if they're clearer